Public/Src/Engine/Processes/SharedOpaqueOutputLogger.cs

```diff
@@ -0,0 +1,233 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.ContractsLight;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using BuildXL.Pips.Operations;
+using BuildXL.Utilities;
+using BuildXL.Utilities.Collections;
+using JetBrains.Annotations;
+
+namespace BuildXL.Processes
+{
+    /// <summary>
+    /// Responsible for keeping a log of all paths written to by a given process pip.
+    /// 
+    /// Paths are flushed to an underlying file as soon as they're reported via <see cref="RecordFileWrite(PathTable, AbsolutePath)"/>.
+    /// 
+    /// A number of root directories may be set in which case the log will record only 
+    /// those paths that fall under one of those directories.  The typical use case is to
+    /// use all shared opaque directory outputs of a process as root directories.
+    /// </summary>
+    /// <remarks>
+    /// NOTE: not thread-safe
+    /// </remarks>
+    public sealed class SharedOpaqueOutputLogger : IDisposable
+    {
+        private readonly HashSet<AbsolutePath> m_recordedPathsCache;
+        private readonly Lazy<BuildXLWriter> m_lazyBxlWriter;
+
+        /// <summary>
+        /// Absolute path of this log file.
+        /// </summary>
+        public string LogPath { get; }
+
+        /// <summary>
+        /// Only paths under these root directories will be recorded by <see cref="RecordFileWrite(PathTable, AbsolutePath)"/>
+        /// 
+        /// When <code>null</code>, all paths are recorded.
+        /// </summary>
+        [CanBeNull]
+        public IReadOnlyList<string> RootDirectories { get; }
+
+        /// <summary>
+        /// Creates a new output logger for a given process.
+        /// 
+        /// Shared opaque directory outputs of <paramref name="process"/> are used as root directories and
+        /// <see cref="GetOutputLogFileForProcess"/> is used as log base name.
+        ///
+        /// <seealso cref="SharedOpaqueOutputLogger(string, IReadOnlyList{string})"/>
+        /// </summary>
+        public SharedOpaqueOutputLogger(PipExecutionContext context, Process process, AbsolutePath logDirectory)

```

Comment:
I believe this class can be made free from `PipExecutionContext`, `Process`, and `PathTable` (as well as `AbsolutePath`). 

The absolute path is only used when you want to check if a path is inside a directory. But in that case, can you not use check for substring?

Reply:
This class is already free of `PipExecutionContext`, `Process`, and `PathTable` (in the sense that it has no fields of any of those types); it's only this constructor that accepts those arguments, and the reason why this constructor exists is just convenience (since this logger will be created per process pip).

If you really don't want to see this constructor here, I can turn it into a static factory method and move it to PipExecutor (or wherever it's currently being called from)